tests/test_updater_delegation_graphs.py

```diff
@@ -251,16 +277,119 @@ def test_graph_traversal(self, test_data: DelegationsTestCase) -> None:
+            exp_calls = [(role, 1) for role in test_data.visited_order]
+
+            updater = self._init_updater()
+            # Call explicitly refresh to simplify the expected_calls list
+            updater.refresh()
+            self.sim.fetch_tracker.metadata.clear()
+
+            with self.assertRaises(UnsignedMetadataError):
+                updater.get_targetinfo("missingpath")
+            # Check that there were no visited roles after the invalid one
+            # and only the valid metadata files were persisted
+            self.assertListEqual(self.sim.fetch_tracker.metadata, exp_calls)
+            self._assert_files_exist(exp_files)
+        finally:
+            self.teardown_subtest()
+
+
+class TestTargetFileSearch(TestDelegations):
+    r"""
+    Create a single repository with the following delegations:
+
+              targets
+         *.doc / \ r/*/*
+              A   B
+            r/x/* / \ r/y/*.zip
+                 C   D
+
+    Test that Updater successfully finds the target files metadata,
+    traversing the delegations as expected.
+    """
+
+    delegations_tree = DelegationsTestCase(
+        delegations=[
+            TestDelegation("targets", "A", paths=["*.doc", "*.md"]),
+            TestDelegation("targets", "B", paths=["releases/*/*"]),
+            TestDelegation("B", "C", paths=["releases/x/*"]),
+            TestDelegation("B", "D", paths=["releases/y/*.zip"]),
+        ],
+        target_files=[
+            TestTarget("targets", b"content", "targetfile"),
+            TestTarget("A", b"content", "README.md"),
+            TestTarget("C", b"content", "releases/x/x_v1"),
+            TestTarget("D", b"content", "releases/y/y_v1.zip"),
+        ],
+    )
+
+    def setUp(self) -> None:
+        super().setUp()
+        self._init_repo(self.delegations_tree)
+
+    targets: utils.DataSet = {
+        "targetfile": ("targetfile", []),
+        "README.md": ("README.md", ["A"]),
+        "releases/x/*": ("releases/x/x_v1", ["B", "C"]),
+        "releases/y/*.zip": ("releases/y/y_v1.zip", ["B", "D"]),
+    }
+
+    @utils.run_sub_tests_with_dataset(targets)
+    def test_targetfile_search(self, test_data: Tuple[str, List[str]]) -> None:
+        try:
+            self.setup_subtest()
+            targetpath, visited_order = test_data
+            exp_files = [*TOP_LEVEL_ROLE_NAMES, *visited_order]
+            exp_calls = [(role, 1) for role in visited_order]
+            updater = self._init_updater()
+            # Call explicitly refresh to simplify the expected_calls list
+            updater.refresh()
+            self.sim.fetch_tracker.metadata.clear()
+            target = updater.get_targetinfo(targetpath)
+            # Confirm that the expected TargetFile is found
+            assert target is not None
+            exp_target = self.sim.target_files[targetpath].target_file
+            self.assertDictEqual(target.to_dict(), exp_target.to_dict())
```

Comment:
since content is the same for every test target, this would return true even if we had two different roles serving the same  targetpath (and we got the wrong target because of a bug)

Reply:
So this is pretty much a bug in the test, I added a different content for each test file.